internal/wrr/random.go

```diff
@@ -87,5 +87,7 @@ func (rw *randomWRR) Add(item any, weight int64) {
 }
 
 func (rw *randomWRR) String() string {
+	rw.mu.Lock()
+	defer rw.mu.Unlock()
```

Comment:
This is risky as we'll deadlock if `String` is ever called while the lock is held.

Note that `String` is called indirectly, too, e.g. `fmt.Sprint(rw)`.  I don't think I like this.  Maybe we should just exclude the `items` from what is printed and print only static info instead, with a different method that can retrieve the detailed data as a string if needed.

Consider:

```go
func (rw *randomWRR) Whatever() {
	rw.mu.Lock()
	defer rw.mu.Unlock()
	...
	if badThing {
		rw.logger.Warningf("saw bad thing in %s", rw)
	}
	// or:
	if rw.logger.V(2) {
		rw.logger.Infof("saw weird thing in %s", rw)
	}
}
```

We'd deadlock in rare cases that might not be covered in tests.

Reply:
I just noticed that `rw.mu` is of type `sync.RWMutex`. Which means i could prolly use `mu.RLock()` in String(), which would avoid deadlocks on the mutex? wdyt? @dfawley

Reply:
Spoke with him offline.. my b, rLock still blocks on wLock.